File: rcsetgenerator.m
#import <Foundation/Foundation.h>

#define bundleID @"com.michael.generator"
#define containerURL [NSURL URLWithString:@"file:///private/var/mobile"]

int main() {
    id enabled = [[[NSUserDefaults alloc] _initWithSuiteName:bundleID container:containerURL] objectForKey:@"enabled"];
    if (enabled != nil) {
        if ([enabled isKindOfClass:[NSNumber class]]) {
            if (![enabled boolValue]) {
                return 0;
            }
        } else {
            [[[NSUserDefaults alloc] _initWithSuiteName:bundleID container:containerURL] removeObjectForKey:@"enabled"];
        }
    }
    execvp("setgenerator", (char *[]){"setgenerator", NULL});
    perror("setgenerator");
    return -1;
}

File: rcsetgenerator.c
#include <CoreFoundation/CoreFoundation.h>

#define bundleID CFSTR("com.michael.generator")

int main() {
    CFArrayRef keyList = CFPreferencesCopyKeyList(bundleID, CFSTR("mobile"), kCFPreferencesAnyHost);
    if (keyList != NULL) {
        if (CFArrayContainsValue(keyList, CFRangeMake(0, CFArrayGetCount(keyList)), CFSTR("enabled")) && !CFBooleanGetValue(CFPreferencesCopyValue(CFSTR("enabled"), bundleID, CFSTR("mobile"), kCFPreferencesAnyHost))) {
            CFRelease(keyList);
            return 0;
        }
        CFRelease(keyList);
    }
    execvp("setgenerator", (char *[]){"setgenerator", NULL});
    perror("setgenerator");
    return -1;
}
